### ISSUE TITLE
Support for @JsonValue custom ID types

### ISSUE THREAD

**Issue by marquiswang** [mongojack/mongojack#224]
I'm finally trying to migrate from 2.x. I have a lot of code of the form:

    JacksonDBCollection.wrap(collection, SomeType.class, CustomId.class, objectMapper);

where `CustomId` is a class with `@JsonValue` and `@JsonCreator` annotations that allow Jackson to serialize it to a string. With 2.x, this was supported and used the String serialization of `CustomId` as the id.

With 3.x and 4.x, the generic key type has been removed from JacksonMongoCollection, and I get the following error when trying to insert:

    java.lang.IllegalArgumentException: Unsupported ID type: class foo.bar.CustomId

Is there any way to get the equivalent old functionality?

**Comment by rlodge** [comment]
Not at this time, at least not without fakery inside the class (like declaring the _id field as String and then converting to/from it...)

**Comment by marquiswang** [comment]
Thanks, that's what I was afraid of. Is this something that could be added?

**Comment by rlodge** [comment]
I _think_ so.  It involves determining if the jackson-converted value will be something mongo will use as an id, though, so the solution will have to be thought through a bit.  I played with some options this morning but I don't know if I like them.

**Comment by marquiswang** [comment]
I managed to make it work by adding

```
try {
    final String stringRepresentation = objectMapper.writer().writeValueAsString(value);
    if (stringRepresentation.startsWith("\"") && stringRepresentation.endsWith("\"")) {
        return new BsonString(stringRepresentation.substring(1, stringRepresentation.length() - 1));
    }
} catch (JsonProcessingException aE) {
    throw new RuntimeException(aE);
}
 ```

at the bottom of `JacksonCodec#constructIdValue`. I don't know if there is a more elegant way to determine if Jackson thinks an object should be serialized as a non-object type (I'm only interested in Strings here, but obviously we could detect numeric types too).

**Comment by marquiswang** [comment]
I'm afraid to report that I've found further code in my codebase (yay legacy codebases) that uses an id field with an object, which apparently mongojack 2.x and Mongo happily allow (totally didn't know that). 

```
> db.foo.bar.findOne()
{
	"_id" : {
		"foo" : "bar",
		"date" : [
			2016,
			4,
			5
		]
	},
	"offset" : 123.456
}
```

If we want MongoJack to support this... that would be really nice (and backwards compatible with 2.x!). Though I could understand if you didn't want to.

(BTW, I found that any version of Jackson below 2.13.3 does not work with Java 17, in case that's what you're running into)

**Comment by marquiswang** [comment]
I found that this works

```
        final JsonNode tree = objectMapper.valueToTree(value);
        if (tree.isDouble()) {
            return new BsonDouble(tree.asDouble());
        } else if (tree.isTextual()) {
            return new BsonString(tree.asText());
        } else if (tree.isInt()) {
            return new BsonInt32(tree.asInt());
        } else if (tree.isLong()) {
            return new BsonInt64(tree.asLong());
        } else if (tree.isBigDecimal()) {
            return new BsonDecimal128(new Decimal128(tree.decimalValue()));
        } else {
            return BsonDocument.parse(tree.toString());
        }
```

**Comment by rlodge** [comment]
I have published a `4.5.1-SNAPSHOT` version that I believe fixes this.  Please let me know if it works for you.

**Comment by rlodge** [comment]
Fix deployed 4.5.1 and 4.7.0

**Comment by marquiswang** [comment]
Thank you!
